cirq/sim/sparse_simulator.py

```diff
@@ -134,34 +135,39 @@ def _run(self, circuit: circuits.Circuit,
-                                                   repetitions)
-        return self._run_sweep_repeat(resolved_circuit, repetitions)
-
-    def _run_sweep_terminal_sample(self, circuit: circuits.Circuit,
-                                   repetitions: int) -> Dict[str, np.ndarray]:
-        for step_result in self._base_iterator(
-                circuit=circuit,
-                qubit_order=ops.QubitOrder.DEFAULT,
-                initial_state=0,
-                perform_measurements=False):
+        qubit_order = sorted(resolved_circuit.all_qubits())
+
+        # Simulate as many unitary operations as possible before having to
+        # repeat work for each sample.
+        unitary_prefix, general_suffix = _split_into_unitary_then_general(
+            resolved_circuit)
+        step_result = None
+        for step_result in self._base_iterator(circuit=unitary_prefix,
+                                               qubit_order=qubit_order,
+                                               initial_state=0,
+                                               perform_measurements=False):
             pass
-        # We can ignore the mixtures since this is a run method which
-        # does not return the state.
-        measurement_ops = [op for _, op, _ in
-                           circuit.findall_operations_with_gate_type(
-                                   ops.MeasurementGate)]
-        return step_result.sample_measurement_ops(measurement_ops,
-                                                  repetitions,
-                                                  seed=self._prng)
-
-    def _run_sweep_repeat(self, circuit: circuits.Circuit,
+        assert step_result is not None
+
+        # When an otherwise unitary circuit ends with non-demolition computation
+        # basis measurements, we can sample the results more efficiently.
+        general_ops = list(general_suffix.all_operations())
+        if all(isinstance(op.gate, ops.MeasurementGate) for op in general_ops):
+            return step_result.sample_measurement_ops(measurement_ops=cast(
+                List[ops.GateOperation], general_ops),
+                                                      repetitions=repetitions,
+                                                      seed=self._prng)
+
+        qid_shape = protocols.qid_shape(qubit_order)
+        intermediate_state = step_result.state_vector().reshape(qid_shape)
+        return self._run_sweep_repeat(initial_state=intermediate_state,
+                                      circuit=general_suffix,
+                                      repetitions=repetitions,
+                                      qubit_order=qubit_order)
+
+    def _run_sweep_repeat(self, initial_state: np.ndarray,
+                          circuit: circuits.Circuit,
+                          qubit_order: 'cirq.QubitOrderOrList',
                           repetitions: int) -> Dict[str, np.ndarray]:
```

Comment:
Is it worth adding a developer-centric docstring?  It's a private method, so probably not mandatory, but might be nice to have.

Reply:
Done. Also renamed it.